sms_statistics_service/sms_statistics_service.py

```diff
@@ -0,0 +1,104 @@
+
+    args = parser.parse_args()
+
+    google_cloud_credentials_file_path = args.google_cloud_credentials_file_path
+    firestore_credentials_url = args.firestore_credentials_url
+    start_minute_inclusive = isoparse(args.start_minute_inclusive)
+    end_minute_exclusive = isoparse(args.end_minute_exclusive)
+
+    assert start_minute_inclusive.second == 0
+    assert start_minute_inclusive.microsecond == 0
+    assert end_minute_exclusive.second == 0
+    assert end_minute_exclusive.microsecond == 0
+
+    log.info("Initialising the Firestore client...")
+    firestore_credentials = json.loads(google_cloud_utils.download_blob_to_string(
+            google_cloud_credentials_file_path, firestore_credentials_url))
+    firestore_client = FirestoreClient(firestore_credentials)
+
+    log.info("Downloading the active project details from Firestore...")
+    active_projects = firestore_client.get_active_projects()
+    log.info(f"Downloaded the details for {len(active_projects)} active projects")
+
+    for project in active_projects:
+        log.info(f"Computing SMS statistics for project {project.project_name}...")
+
+        log.info(f"Downloading the Rapid Pro credentials token from '{project.rapid_pro_token_url}'...")
+        rapid_pro_token = google_cloud_utils.download_blob_to_string(
+            google_cloud_credentials_file_path, project.rapid_pro_token_url).strip()
+
+        log.info(f"Downloading raw messages from Rapid Pro...")
+        rapid_pro = RapidProClient(project.rapid_pro_domain, rapid_pro_token)
+        raw_messages = rapid_pro.get_raw_messages(created_after_inclusive=start_minute_inclusive,
+                                                  created_before_exclusive=end_minute_exclusive)
+
+        log.info("Computing message stats for each minute...")
+        # Create a table of counts for all the minutes of interest, with all counts initialised to 0
+        stats = dict()  # of minute iso_string -> SMSStats
+        minute = start_minute_inclusive
+        while minute < end_minute_exclusive:
+            stats[minute.astimezone(pytz.utc).isoformat(timespec="minutes")] = SMSStats()
+            minute += datetime.timedelta(minutes=1)
+
+        # Loop over all of the downloaded messages and increment the appropriate count
+        unhandled_status_count = 0
+        for msg in raw_messages:
+            minute_stats = stats[msg.created_on.astimezone(pytz.utc).isoformat(timespec="minutes")]
+
+            # Message statuses are "documented" here:
+            # https://github.com/rapidpro/rapidpro/blob/c972205aae29f7219582fc29478e8ecacb579f9f/temba/msgs/models.py#L79
+            if msg.direction == "in":
+                minute_stats.total_received += 1
+            elif msg.direction == "out":
+                if msg.status in {"initializing", "pending", "queued"}:
+                    pass  # Rapid Pro has not attempted to actually send the message yet, so don't increment any totals
```

Comment:
I think I'd like these added to a counter. If we see a larger number of them it's a good warning that something is going wrong with the instance

Reply:
changed to minute_stats.total_pending += 1